### PR TITLE
[wip] Enable LPC support in modm

### PR DESCRIPTION
**Pull request by strongly-typed**
This is a WIP of enabling LPC support (again) in modm and basis for discussion.

Tested with LPCxpresso LPC11C24 board

- [x] GPIO
- [x] UART Tx
- [x] CAN Rx and Tx
- [x] ADC Channel 0 and Channel 1

### CODE REVIEW

File: src/modm/platform/can/lpc/module.lb [hunk 1]
```diff
@@ -33,7 +33,12 @@ def prepare(module, options):
             maximum=256,	# FIXME
             default=16))
 
-    module.depends(":platform:core", ":debug", ":platform:clock.generic")
+    module.depends(
+        ":architecture:can",
+        ":platform:core",
```
Comment (salkinium) [inline src/modm/platform/can/lpc/module.lb]: Don't depend on `:platform:core`, only on the actual interfaces you need from `:architecture:*`.

File: src/modm/platform/clock/lpc/clock.cpp.in [hunk 2]
```diff
@@ -13,6 +14,13 @@
 #include "../device.hpp"
 #include "clock.hpp"
 
+namespace modm::clock
+{
+uint32_t modm_fastdata fcpu({{ "{0:,}".format(irc_frequency_mhz|int * 1000000).replace(',', "'") }});
+uint32_t modm_fastdata fcpu_kHz({{ "{0:,}".format(irc_frequency_mhz|int * 1000).replace(',', "'") }});
+uint16_t modm_fastdata fcpu_MHz({{ irc_frequency_mhz }});
+uint16_t modm_fastdata ns_per_loop({{ "{0:,}".format((loops * 1000.0 / irc_frequency_mhz)|int).replace(',', "'") }});
```
Comment (salkinium) [inline src/modm/platform/clock/lpc/clock.cpp.in:22]: What is this formatting monstrosity? 😳

File: src/modm/platform/clock/lpc/clock_impl.hpp.in [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+template< uint32_t Core_Hz, uint16_t Core_mV = 3'300 >
+uint32_t
+ClockControl::setFlashLatency()
+{
+	// constexpr flash_latency fl = computeFlashLatency(Core_Hz, Core_mV);
+	// static_assert(Core_Hz <= fl.max_frequency, "CPU Frequency is too high for this core voltage!");
+
+	if (Core_Hz <= 20_MHz) {
```
Comment (salkinium) [inline src/modm/platform/clock/lpc/clock_impl.hpp.in]: use `if constexpr`.

File: src/modm/platform/clock/lpc/module.lb [hunk 4]
```diff
@@ -11,31 +11,54 @@
...
 
 def prepare(module, options):
     if not options[":target"].has_driver("clock:lpc"):
+        print('lpc_clock FALSE')
         return False
 
-    module.depends(":platform:core")
+    # module.depends(":platform:core")
+    module.depends(":cmsis:device", ":utils", ":platform:clock")
+    print('lpc_clock TRUEE')
```
Comment (salkinium) [inline src/modm/platform/clock/lpc/module.lb]: \*hicks\* that… that is… so TRUEEEEE \*hicks\* 🥴 \*drinks more wine\*

File: src/modm/platform/clock/lpc/module.lb [hunk 5]
```diff
@@ -11,31 +11,54 @@
...
     properties["target"] = target = device.identifier
+    properties["partname"] = device.partname
+    properties["core"] = core = device.get_driver("core")["type"]
     properties["driver"] = driver
 
+    if target["family"] in ["11"]:
+        properties["irc_frequency_mhz"] = 12 # MHz
+    else:
+        print("Internal RC oscillator frequency unknown.")
+        raise Exception("Internal RC oscillator frequency yet unknown for family {}".format(family))
```
Comment (salkinium) [inline src/modm/platform/clock/lpc/module.lb]: Throwing `ValidateException`s should happen in the `def validate(env):` step, otherwise there is a partial build of the library which isn't very elegant.

File: src/modm/platform/clock/lpc/static.hpp.in [hunk 6]
```diff
@@ -1,91 +1,109 @@
...
+ * THIS SOFTWARE IS PROVIDED BY ROBOTERCLUB AACHEN E.V. ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ROBOTERCLUB AACHEN E.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Comment (salkinium) [inline src/modm/platform/clock/lpc/static.hpp.in]: wrong license?
And this entire file needs to die.

File: src/modm/platform/clock/lpc/static.macros [hunk 7]
```diff
@@ -55,18 +55,6 @@ public:
...
-		Rcc::setApb1Prescaler(Rcc::Apb1Prescaler::Div4);
-		Rcc::setApb2Prescaler(Rcc::Apb2Prescaler::Div2);
-%% elif partname.startswith("stm32f100")
-		Rcc::setApb1Prescaler(Rcc::Apb1Prescaler::Div1);
-		Rcc::setApb2Prescaler(Rcc::Apb2Prescaler::Div1);
-%% elif target["platform"] == "stm32" and target["family"] in ["f1", "f3"]
-		Rcc::setApb1Prescaler(Rcc::Apb1Prescaler::Div2);
-		Rcc::setApb2Prescaler(Rcc::Apb2Prescaler::Div1);
-%% endif
-
```
Comment (salkinium) [inline src/modm/platform/clock/lpc/static.macros:69]: kill this code with :fire:.

File: src/modm/platform/core/lpc/module.lb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    module.name = ":platform:core"
+    module.description = """\
+# NXP LPC core module
+
+Provides LPC specific linkerscripts and startup code.
+"""
+
+def prepare(module, options):
+    if options[":target"].identifier.platform != "lpc":
+        print('module.lb platform clock lpc FALSE')
```
Comment (salkinium) [inline src/modm/platform/core/lpc/module.lb:23]: `lbuild -vv build` goes a long way you know.

File: src/modm/platform/core/lpc/startup_platform.c.in [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+ * initialized.
+ * In the worst case you won't even have access to the stack, if the
+ * memory containing the stack is not physically enabled yet.
+ * In that case, consider using inline assembly to manage stack access
+ * manually, until the memory is enabled.
+ */
+void
+__modm_initialize_platform(void)
+{
+	modm_gpio_enable();
```
Comment (salkinium) [inline src/modm/platform/core/lpc/startup_platform.c.in:27]: This function needs to be pointed into the `.hardware_init` section, so that there is no linker requirement from the `:platform:core` to `:platform:gpio`. The `:platform:gpio` module should be strictly optional.

Add this code to whatever file the `modm_gpio_enable` function is defined.
```c++
// .A000100 postfix since .hardware_init is sorted alphabetically and
// this should be executed before application inits.
modm_section(".hardware_init.A000100_modm_gpio_enable")
uint32_t modm_gpio_enable_ptr = (uint32_t) &modm_gpio_enable;
```

File: src/modm/platform/gpio/lpc/module.lb [hunk 10]
```diff
@@ -11,26 +11,130 @@
...
+    module.description = "General Purpose I/O (GPIO) of NXP LPC devices"
 
 def prepare(module, options):
-    if not options[":target"].has_driver("gpio:lpc"):
+    device = options[":target"]
+    if not device.has_driver("gpio:lpc"):
         return False
 
-    module.depends(":platform:core", ":platform:gpio.generic", ":platform:i2c.generic")
+    # ToDo: Derive from device
```
Comment (salkinium) [inline src/modm/platform/gpio/lpc/module.lb:59]:
```diff
-    # ToDo: Derive from device
+    # TODO: Derive from device
```
I mean… that's like a BIG todo… 🤪

File: src/modm/platform/gpio/lpc/pin.hpp.in [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+};
+
+/// IO class for Pin {{port }}_{{ pin}}
+/// @ingroup	modm_platform_gpio
+class Gpio{{ port }}_{{ pin }} : public Gpio, public ::modm::GpioIO
+{
+	// template<class... Gpios>
+	// friend class GpioSet;
+	// using PinSet = GpioSet<Gpio{{ port ~ pin }}>;
+	using PinSet = GpioSet{{ port }}_{{ pin }};
```
Comment (salkinium) [inline src/modm/platform/gpio/lpc/pin.hpp.in:114]: Well… GpioSet should be done right.

File: src/modm/platform/gpio/stm32/module.lb [hunk 12]
```diff
@@ -237,6 +240,10 @@ def validate(env):
         all_peripherals.extend([get_driver(r) for r in driver["remap"]])
     bprops["all_peripherals"] = sorted(list(set(all_peripherals)))
     bprops["all_signals"] = sorted(list(set(s["name"] for s in all_signals.values())))
+
+    pp(bprops["all_peripherals"])
+    pp(bprops["all_signals"])
+
```
Comment (salkinium) [inline src/modm/platform/gpio/stm32/module.lb:246]: modm-devices needs a much better interface for accessing the raw data and derived data. The data format is optimized for unambiguity, not accessiblity.

File: tools/build_script_generator/openocd.cfg.in [hunk 13]
```diff
@@ -14,7 +14,7 @@ proc modm_itm_log { OUTPUT F_CPU {BAUDRATE ""} } {
 }
 
 proc modm_program { SOURCE } {
-	program $SOURCE verify
+	program $SOURCE
```
Comment (salkinium) [inline tools/build_script_generator/openocd.cfg.in:21]: Did you break OpenOCD again? 🤨

File: tools/build_script_generator/scons/resources/SConscript.in [hunk 14]
```diff
@@ -115,7 +115,7 @@ env["CONFIG_AVRDUDE_OPTIONS"] = "{{ avrdude_options }}"
 %% if avrdude_baudrate > 0
 env["CONFIG_AVRDUDE_BAUDRATE"] = "{{ avrdude_baudrate }}"
 %% endif
-%% elif platform in ["stm32"]
+%% elif platform in ["stm32"] or platform in ["lpc"]
```
Comment (salkinium) [inline tools/build_script_generator/scons/resources/SConscript.in:118]: this should probably check for `core.startwith("cortex-m")`.

File: tools/devices/lpc/lpc111.xml [hunk 15]
```diff
@@ -1,103 +0,0 @@
...
-      <gpio port="2" pin="11" func="Gpio,Spi1,Timer32_0"         bits="mode,hys,od"/>
-      <gpio port="3" pin="0"  func="Gpio,Uart,Timer16_0"         bits="mode,hys,od"/>
-      <gpio port="3" pin="1"  func="Gpio,Uart,Timer16_0"         bits="mode,hys,od"/>
-      <gpio port="3" pin="2"  func="Gpio,Uart,Spi1,Timer16_0"    bits="mode,hys,od"/>
-      <gpio port="3" pin="3"  func="Gpio,Uart,Timer16_0"         bits="mode,hys,od"/>
-      <gpio port="3" pin="4"  func="Gpio,Uart,Timer16_0"         bits="mode,hys,od"/>
-      <gpio port="3" pin="5"  func="Gpio,Uart,Timer16_0"         bits="mode,hys,od"/>
-    </driver>
-  </device>
-</modm>
```
Comment (salkinium) [inline tools/devices/lpc/lpc111.xml:103]: Why were these moved to modm-devices? They are not at the level of quality or automation that warrants them being there.

File: src/modm/platform/uart/lpc/module.lb [hunk 16]
```diff
@@ -59,21 +57,36 @@ class Instance(Module):
...
+    module.description = "Universal Asynchronous Receiver Transmitter (UART)"
 
 def prepare(module, options):
     device = options[":target"]
     if not device.has_driver("uart:lpc"):
         return False
 
-    module.depends(":platform:core")
+    module.depends(
+        ":platform:core",
```
Comment (salkinium) [inline src/modm/platform/uart/lpc/module.lb]: no

File: src/modm/platform/gpio/lpc/unused.hpp.in [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+	static void setInputTrigger(const InputTrigger) {}
+	static bool getExternalInterruptFlag() { return false; }
+	/// Reset the interrupt flag in the interrupt routine.
+	static void acknowledgeExternalInterruptFlag() {}
+
+	// GpioIO
+	// start documentation inherited
+	static Direction getDirection() { return Direction::Special; }
+	// end documentation inherited
+	static void lock() {}
```
Comment (salkinium) [inline src/modm/platform/gpio/lpc/unused.hpp.in:109]: This needs to have the same API as the platform specific GPIO classes. It should be drop-in replacable.

File: src/modm/board/lpcxpresso_lpc11c24/module.lb [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+http://www.embeddedartists.com/products/lpcxpresso/lpc11C24_xpr.php
+"""
+
+def prepare(module, options):
+    if options[":target"].partname != "lpc11c24":
+        return False
+
+    module.depends(
+        ":platform:lpc",
+        ":platform:core",
```
Comment (salkinium) [inline src/modm/board/lpcxpresso_lpc11c24/module.lb]: no

File: .circleci/config.yml [hunk 19]
```diff
@@ -111,6 +111,17 @@ jobs:
...
+    docker:
+      - image: modm/modm-build:latest
+    steps:
+      - checkout
+      - run: (git submodule sync && git submodule update --init --jobs 8) & pip3 install -U lbuild & wait
+      - run:
+          name: Examples NXP LPC11 Series
+          command: |
+            (cd examples && ../tools/scripts/examples_compile.py lpcxpresso)
+
```
Comment (salkinium) [inline .circleci/config.yml:124]: It would be nice to only have 4 initial jobs (as CircleCI only gives us 4 parallel jobs), so they all start and finish within 3mins or so.

**Review by salkinium — COMMENTED** [review]
Hm… I think the biggest issue with with the device files and whether they can be of similar quality of the STM32 or SAM files. Furthermore, the CMSIS-Header files for LPC should be moved out of ext/ and into their own repository, with auto-updates similar to cmsis-header-stm32.

File: src/modm/platform/clock/lpc/clock.cpp.in (hunk 2 above)
Comment (strongly-typed) [reply src/modm/platform/clock/lpc/clock.cpp.in:22]: [See here](https://github.com/modm-io/modm/blob/develop/src/modm/platform/clock/stm32/rcc.cpp.in#L23). Other suggestions?

Comment (salkinium) [reply src/modm/platform/clock/lpc/clock.cpp.in:22]: Oh, because of the `'` in the numbers. Fine. A little overkill, but that's what we do here…

**Comment by salkinium** [comment]
I’m closing this, since I don’t see broad LPC support happening without significant effort in modm-devices first. Your changes will be archived in this PR so they’ll still be available for future efforts.
